ui-speedspacechart/src/components/helpers/drawElements/speedLimitTags.ts

```diff
@@ -82,108 +86,132 @@ export const drawSpeedLimitTags = async ({
-            ICON_WIDTH,
-            ICON_HEIGHT,
-            '#FFFFFF'
-          );
-
-          if (tag === 'incompatible' && cursor.x && cursor.y) {
-            if (
-              cursor.x >= x - MARGIN_LEFT + leftOffset &&
-              cursor.x <= x - MARGIN_LEFT + leftOffset + tagWidth &&
-              cursor.y >= marginTop - MARGIN_TOP + Y_POSITION - 2 &&
-              cursor.y <= marginTop - MARGIN_TOP + Y_POSITION - 1 + RECT_HEIGHT
-            ) {
-              tooltip = {
-                cursorX: cursor.x + MARGIN_LEFT,
-                cursorY: Y_POSITION,
-                text: 'Incompatible with the infrastructure',
-              };
-            }
-          }
-        } else {
-          ctx.fillStyle = 'white';
-          ctx.font = '600 12px "IBM Plex Sans"';
-          ctx.textAlign = 'left';
-          ctx.textBaseline = 'middle';
-          const textPosition = x + TEXT_LEFT_PADDING;
-          ctx.fillText(tag, textPosition, Y_POSITION + TEXT_PADDING_TOP + RECT_HEIGHT / 2);
-        }
+      if (tag === 'incompatible' || tag === 'missing_from_train') {
+        const image = tag === 'incompatible' ? alertFillImage : questionImage;
+
+        ctx.fillStyle = color;
+
+        const iconXPosition = x + (tagWidth - ICON_BACKGROUND_WIDTH) / 2;
+        const iconYPosition = Y_POSITION - ICON_OFFSET;
+        const cornerRadius = 4;
+
+        drawRoundedRect(
+          ctx,
+          iconXPosition,
+          iconYPosition,
+          ICON_BACKGROUND_WIDTH,
+          ICON_BACKGROUND_HEIGHT,
+          cornerRadius
+        );
+
+        drawSvgImageWithColor(
+          ctx,
+          image,
+          iconXPosition + ICON_OFFSET,
+          iconYPosition + ICON_OFFSET,
+          ICON_WIDTH,
+          ICON_HEIGHT,
+          WHITE.hex()
+        );
+      } else {
+        ctx.fillStyle = 'white';
+        ctx.font = '600 12px "IBM Plex Sans"';
+        ctx.textAlign = 'left';
+        ctx.textBaseline = 'middle';
+        const textPosition = x + TEXT_LEFT_PADDING;
+        ctx.fillText(tag, textPosition, Y_POSITION + TEXT_PADDING_TOP + RECT_HEIGHT / 2);
       }
-    });
+    }
+  });
+
+  drawSeparatorLinearLayer(ctx, 'rgba(0,0,0,0.1)', MARGINS, width, marginTop);
+  ctx.restore();
+
+  // prevent overlapping with margins left and right
+  ctx.clearRect(0, 0, MARGIN_LEFT, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+  ctx.clearRect(width - MARGIN_RIGHT, 0, width, LINEAR_LAYERS_HEIGHTS.SPEED_LIMIT_TAGS_HEIGHT);
+};
+
+export const computeTooltip = async ({
+  width,
+  height: marginTop,
+  store,
+}: DrawFunctionParams): Promise<tooltipInfos | null> => {
+  const { speedLimitTags, ratioX, leftOffset, cursor } = store;
+
+  const { MARGIN_TOP, MARGIN_LEFT } = MARGINS;
+
+  const maxPosition = maxPositionValue(store.speeds);
+
+  if (!speedLimitTags) return null;
+
+  for (const { position, value } of speedLimitTags) {
+    const tooltipTextMap: Record<string, string> = {
```

Comment:
Sounds a bit weird to have an object with just a single entry, and I don't think an explicit type is required here, but I don't mind too much.

Reply:
I'll confess I didn't look at the tooltip code at all, but I'll do

Reply:
Ah no I thought you were talking about something else but I remember now

Due to how the code was structured I assumed there might be tooltips added for other tags in the future, and I wanted for their addition to be as smooth as possible, without nested if conditions like were present in the previous version of the code.

And some type indication is needed for potentialTooltipText to later know it is a string (even though I agree the ts engine should have been able to deduce that by itself)